asb-native/src/main/java/org/ballerinax/asb/util/ASBUtils.java

```diff
@@ -222,4 +248,164 @@ public static AmqpRetryOptions getRetryOptions(BMap<BString, Object> retryConfig
+        LOGGER.debug("Type of java object of received message " + messageId + " is " + jValue.getClass());
+        Class<?> clazz = jValue.getClass();
+        switch (clazz.getSimpleName()) {
+            case "Integer":
+            case "Long":
+            case "Float":
+            case "Double":
+            case "Boolean":
+            case "Byte":
+            case "Short":
+            case "Character":
+                return Optional.of(jValue);
+            case "String":
+                return Optional.of(StringUtils.fromString((String) jValue));
+            case "BigDecimal":
+                return Optional.of(ValueCreator.createDecimalValue((BigDecimal) jValue));
+            default:
+                LOGGER.debug("java object with type '" + clazz + "' can not be converted to as a Ballerina value");
+                return Optional.empty();
+        }
+    }
+
+    /**
+     * Converts `byte[]` value to the intended Ballerina type.
+     *
+     * @param type  expected type
+     * @param value Value to be converted
+     * @return value with the intended type
+     */
+    public static Object getValueWithIntendedType(byte[] value, Type type) {
+        String strValue = new String(value, StandardCharsets.UTF_8);
+        Object intendedValue = null;
+        try {
+            switch (type.getTag()) {
+                case STRING_TAG:
+                    intendedValue = StringUtils.fromString(strValue);
+                    break;
+                case XML_TAG:
+                    intendedValue = XmlUtils.parse(strValue);
+                    break;
+                case ANYDATA_TAG:
+                    intendedValue = ValueCreator.createArrayValue(value);
+                    break;
+                case RECORD_TYPE_TAG:
+                    intendedValue = CloneWithType.convert(type, JsonUtils.parse(strValue));
+                    break;
+                case UNION_TAG:
+                    if (isSupportedUnionType(type)) {
+                        intendedValue = getValueWithIntendedType(value, getExpectedTypeFromNilableType(type));
+                    } else {
+                        intendedValue = ErrorCreator.createError(StringUtils.fromString("Union types are not " +
+                                "supported for the contextually expected type, except for nilable types"));
+                    }
+                    break;
+                case ARRAY_TAG:
+                    if (getReferredType(((ArrayType) type).getElementType()).getTag() == BYTE_TAG) {
+                        intendedValue = ValueCreator.createArrayValue(value);
+                        break;
+                    }
+                    /*-fallthrough*/
+                default:
+                    intendedValue = getValueFromJson(type, strValue);
+            }
+        } catch (BError bError) {
+            intendedValue = bError;
+        }
+        if (intendedValue instanceof BError) {
```

Comment:
What should we do to handle this case?

Reply:
@abeykoon We should ideally introduce asb specific error subtypes (`asb:Error` and its subtypes) for better error handling. Thought of doing error handling improvements in a separate PR, which will also make it easier to review.

Reply:
+1